File: src-tauri/apptracker_main.py
import time
import psutil
import re
import logging
from flask import Flask, jsonify
from threading import Thread
from PIL import Image
import pygetwindow as gw
import win32process
import win32gui
from titlecase import titlecase
from flask_cors import CORS

class ScreenActivityTracker:
    def __init__(self):
        self.process_time = {}
        self.timestamp = {}
        self.do_run = True

    def get_active_window_title(self):
        try:
            active_window = gw.getActiveWindow()
            return active_window.title
        except Exception as e:
            logging.error(f"Error getting active window title: {e}")
            return None

    def track_activity(self, sleep_time):
        '''Track the active window and the time spent on it.'''
        while self.do_run:
            try:
                current_app = psutil.Process(win32process.GetWindowThreadProcessId(win32gui.GetForegroundWindow())[1]).name().replace(".exe", " ")
                current_app = current_app.replace('ApplicationFrameHost', '')
                current_window_title = self.get_active_window_title()

                self.timestamp[current_app + current_window_title] = int(time.time())
                time.sleep(sleep_time)
                if current_app + current_window_title not in self.process_time:
                    self.process_time[current_app + current_window_title] = 0

                self.process_time[current_app + current_window_title] += int(time.time()) - self.timestamp[current_app + current_window_title]

                parts = re.split(' - | — ', current_window_title)
                for part in parts:
                    part = titlecase(part)
                    if part not in self.process_time:
                        self.process_time[part] = 0
                    self.process_time[part] += int(time.time()) - self.timestamp[current_app + current_window_title]
            except Exception as e:
                logging.error(f"Error tracking activity: {e}")

    def get_sorted_process_time(self):
        '''Return the process time in descending order.'''
        sorted_process_time = {k: v for k, v in sorted(self.process_time.items(), key=lambda item: item[1], reverse=True)}
        return jsonify(sorted_process_time)

def create_app(tracker):
    '''Create a Flask app to serve the process time data.'''
    app = Flask(__name__)
    CORS(app)  # Enable CORS for all routes

    @app.route('/data', methods=['GET'])
    def get_data():
        return tracker.get_sorted_process_time()

    return app

if __name__ == "__main__":
    import argparse 
    # Setting up sleep time and port number to the argument
    parser = argparse.ArgumentParser(description='Track screen activity.')
    parser.add_argument('--sleep_time', type=int, default=1, help='The time to sleep between checks for the active window.')
    parser.add_argument('--port', type=int, default=5000, help='The port to run the Flask app on.')
    args = parser.parse_args()

    # logging.basicConfig(filename='screen_activity_tracker.log', level=logging.INFO)

    # Create a ScreenActivityTracker instance 
    tracker = ScreenActivityTracker()
    tracking_thread = Thread(target=tracker.track_activity, args=(args.sleep_time,))
    tracking_thread.daemon = True
    tracking_thread.start()

    app = create_app(tracker)
    app.run(port=args.port)

File: src-tauri/apptracker_main.js
// activity-tracker.js

const { app, BrowserWindow } = require('electron');
const activeWindow = require('active-win');
const express = require('express');
const cors = require('cors');

class ScreenActivityTracker {
  constructor() {
    this.processTime = {};
    this.timestamp = {};
    this.doRun = true;
  }

  async getActiveWindowTitle() {
    try {
      const result = await activeWindow();
      return result ? result.title : null;
    } catch (e) {
      console.error(`Error getting active window title: ${e}`);
      return null;
    }
  }

  async trackActivity(sleepTime) {
    while (this.doRun) {
      try {
        const activeWindow = await this.getActiveWindowTitle();
        if (activeWindow) {
          const currentTimestamp = Math.floor(Date.now() / 1000);
          
          if (!(activeWindow in this.processTime)) {
            this.processTime[activeWindow] = 0;
          }
          
          if (activeWindow in this.timestamp) {
            this.processTime[activeWindow] += currentTimestamp - this.timestamp[activeWindow];
          }
          
          this.timestamp[activeWindow] = currentTimestamp;
          
          // Split window title and track individual parts
          const parts = activeWindow.split(/\s-\s|\s—\s/);
          for (const part of parts) {
            const titleCasePart = part.replace(/\w\S*/g, (txt) => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase());
            if (!(titleCasePart in this.processTime)) {
              this.processTime[titleCasePart] = 0;
            }
            this.processTime[titleCasePart] += currentTimestamp - this.timestamp[activeWindow];
          }
        }
        
        await new Promise(resolve => setTimeout(resolve, sleepTime * 1000));
      } catch (e) {
        console.error(`Error tracking activity: ${e}`);
      }
    }
  }

  getSortedProcessTime() {
    return Object.fromEntries(
      Object.entries(this.processTime).sort(([,a],[,b]) => b-a)
    );
  }
}

function createApp(tracker) {
  const app = express();
  app.use(cors());

  app.get('/data', (req, res) => {
    res.json(tracker.getSortedProcessTime());
  });

  return app;
}

function main() {
  const tracker = new ScreenActivityTracker();
  const expressApp = createApp(tracker);

  // Start tracking in the background
  tracker.trackActivity(1);

  // Start express server
  const port = 5000;
  expressApp.listen(port, () => {
    console.log(`Server running on http://localhost:${port}`);
  });

  // Create Electron window
  app.whenReady().then(() => {
    const win = new BrowserWindow({
      width: 800,
      height: 600,
      webPreferences: {
        nodeIntegration: true,
        contextIsolation: false
      }
    });

    win.loadURL('http://localhost:5173'); // Assuming your Svelte app runs on this port
  });
}

main();